Documentation/project-docs/developer-guide.md

```diff
@@ -16,4 +11,235 @@ The CoreFX repo can be built from a regular, non-admin command prompt. The build
+
+You can build the DLL for System.Diagnostics.DiagnosticSource.dll by going to the src\System.Diagnostics.DiagnosticsSource\src directory and typing `msbuild System.Diagnostics.DiagnosticSource.builds`. The DLL ends up as  bin\AnyOS.AnyCPU.Debug\System.Diagnostics.DiagnosticSource\System.DiagnosticSource.dll.
+
+You can build the tests for System.Diagnostics.DiagnosticSource.dll by going to 
+src\System.Diagnostics.DiagnosticSource\tests and typing `msbuild System.Diagnostics.DiagnosticSource.Tests.builds`.
+
+There is also a pkg directory for each project, and if you go into it and type `msbuild`, it will build the DLL (if needed)
+and then also build the NuGet package for it. The NuGet package ends up in the bin\packages directory.  
+
+### Building other OSes
+
+By default, building from the root will only build the libraries for the OS you are running on. One can
+build for another OS by specifying `build-managed -FilterToOSGroup=[value]` or build for all by specifying `build-managed -BuildAllOSGroups`.
+
+### Building in Release or Debug
+
+By default, building from the root or within a project will build the libraries in Debug mode. 
+One can build in Debug or Release mode from the root by doing `build -release` or `build -debug` or when building a project by specifying `/p:ConfigurationGroup=[Debug|Release]` after the `msbuild` command.
+
+### Building other Architectures
+
+One can build 32 or 64 bit binaries or for any architecture by specifying in the root `build -buildArch=[value]` or in a project `/p:Platform=[value]` after the `msbuild` command.
+
+
+### Tests
+
+We use the OSS testing framework [xunit](http://xunit.github.io/).
+
+#### Running tests on the command line
+
+By default, the core tests are run as part of the build. Running the tests from
+the command line is as simple as invoking `build.cmd` on windows, and `run-test.sh` on linux and osx.
+
+You can also run the tests for an individual project by building it individually, e.g.:
+
+```
+cd src\System.Collections.Immutable\tests
+msbuild /t:BuildAndTest (or /t:Test to just run the tests if the binaries are already built)
+```
+
+It is possible to pass parameters to the underlying xunit runner via the `XunitOptions` parameter, e.g.:
+```cmd
+msbuild /t:Test "/p:XunitOptions=-class Test.ClassUnderTests"
+```
+
+There may be multiple projects in some directories so you may need to specify the path to a specific test project to get it to build and run the tests.
+
+Tests participate in the incremental build.  This means that if tests have already been run, and inputs to the incremental build have not changed, rerunning the tests target will not execute the test runner again.  To force re-executing tests in this situation, use `msbuild /t:clean;build;test`.
```

Comment:
Actually here is the place to mention it. I think we should say `/t:RebuildAndTest` instead of individually listing the targets here, although that is definitely a valid option also.